server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -1124,48 +1129,59 @@ private WildcardExpressionResolver() {
-                        expand(context, dataStreamsAbstractions, expressions.isEmpty() ? "_all" : expressions.get(0))
-                    );
-                    return new ArrayList<>(resolvedIncludingDataStreams);
-                }
+        /**
+         * Returns all the indices and all the datastreams, considering the open/closed, system, and hidden context parameters.
+         * Depending on the context, returns the names of the datastreams themselves or their backing indices.
+         */
+        private static Collection<String> innerResolveAll(Context context) {
+            List<String> resolvedExpressions = resolveEmptyOrTrivialWildcard(context);
+            if (context.includeDataStreams() == false) {
                 return resolvedExpressions;
+            } else {
+                final List<IndexAbstraction> dataStreamsAbstractions = context.getState()
+                    .metadata()
+                    .getIndicesLookup()
+                    .values()
+                    .stream()
+                    .filter(indexAbstraction -> indexAbstraction.getType() == IndexAbstraction.Type.DATA_STREAM)
+                    .collect(Collectors.toList());
+                // dedup backing indices if expand hidden indices option is true
+                Set<String> resolvedIncludingDataStreams = new HashSet<>(resolvedExpressions);
+                expandMatches(context, dataStreamsAbstractions, false, resolvedIncludingDataStreams::add);
+                return resolvedIncludingDataStreams;
             }
-
-            Set<String> result = innerResolve(context, expressions);
-
-            if (result == null) {
-                return expressions;
-            }
-            if (result.isEmpty() && context.getOptions().allowNoIndices() == false) {
-                IndexNotFoundException infe = new IndexNotFoundException((String) null);
-                infe.setResources("index_or_alias", expressions.toArray(new String[0]));
-                throw infe;
-            }
-            return new ArrayList<>(result);
         }
 
-        private static Set<String> innerResolve(Context context, List<String> expressions) {
-            Set<String> result = null;
+        /**
+         * Returns all the existing resource (index, alias and datastream) names that the {@param expressions} resolve to.
+         * The passed-in {@param expressions} might contain wildcards and exclusions, as well as plain resource names.
+         * The return is a {@code Collection} (usually a {@code Set} but can also be a {@code List}, for performance reasons) of plain
+         * resource names only. All the returned resources are "accessible", in the given context, i.e. the resources exist
+         * and are not an alias or a datastream if the context does not permit it.
+         * Wildcard expressions, depending on the context:
+         *   - might not cover hidden or system resources (but plain names can refer to hidden or system resources)
```

Comment:
nit: might not "cover" .. cover == resolve, right ? if so I think it would be a bit more clear to use "resolve" as above.

Reply:
I have pushed https://github.com/elastic/elasticsearch/pull/89311/commits/6b3f9684e4a775a046c41a68258562b26a99fcff